### ISSUE TITLE
The `final` keyword is too long

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#136]
It's inconvenient that declaring final variables require more typing than non-final variables. This discourages the use of final variables.

Even with inferred types, it's shorter, and if you want to specify the type, it gets even longer:
```dart
var foo = ...;
String foo = ...;
final foo = ...;
final String foo = ...;
```

While final variables are not strictly necessary (if you don't assign to a variable, it doesn't matter semantically whether it's declared final or not), some people prefer a style where you make variables final unless they need to be mutable, and Dart does not support that writing style very well.

See #135 for one suggestion which improves the experience for inferred variables by replacing `final` with `val`. It doesn't improve the typed case. It requires adding a new built-in identifier (but likely not a reserved word).

Another option could be using `:=` for final initialization (since final local variables need to be initialized):
```dart:
foo := ...;
String foo := ...;
```
This doesn't work for instance variables since they may be intialized by a constructor, or parameters, and to keep requiring the `final` word for instance variables and parameters, and using `:=` everywhere else, is inconsistent.

**Comment by peldritch** [comment]
`let` could be preferable to `val` or `:=` for immutable bindings:

`val` implies the rhs is a value object Kotlin style, not just a binding.
`val` is tricky to distinguish from `var` when reading/eyeballing code.
`let` is used in kernel
`let` is more pleasant to touch-type on english keyboards than `val`
`:=` is less touch-type friendly
`let result = when { } ` is aesthetic for expression/functional style programming.

**Comment by zoechi** [comment]
I like `final` because it stands out. 
The distinction between `val` and `var` is easy to overlook, 
but it's easy to switch between `var` and `val`.
Usually readability should have higher priority.

**Comment by yjbanov** [comment]
I think a bigger problem with `final` is that it's not the default. I'd rather have to put `var` to mean that I want something to be, well, _variable_, and in the case of local variables I'd want the compiler to scream at me if I don't actually mutate it. So:

```dart
String foo = readFoo(); // final by default
final foo = readFoo(); // same thing but infers type
var String foo = readFoo(); // foo is mutable
var foo = readFoo();  // type inference
```

This would also solve @pschiffmann's problem with function parameters.

Also, I agree with @zoechi that `let` is a better keyword for this.

**Comment by mraleph** [comment]
Note that `let` is a non-final binding in JavaScript so it can be confusing for people coming from JS.

**Comment by peldritch** [comment]
Dart is enough of a distinct entity to JavaScript that it's usage re `let` can differ.

Also, JavaScript devs may appreciate a succinct `let` vs a more verbose and misleading `const` for non-final bindings.

**Comment by peldritch** [comment]
`let` comes from the functional world and functional concepts are gaining dev mindshare.

So `let` could be part of a story re introducing more functional and expression orientated elements into Dart.

**Comment by Hixie** [comment]
It'd be neat if you could opt-in to making `final` the default.

`{$FINALDEFAULT ON}`
or
`#pragma final_default on`
or
`.ALLFINAL`
or
`use final;`
or
`from __future__ import everything_final`

**Comment by kasperpeulen** [comment]
Swift also uses `let` and has the same meaning as `final` in Dart.

**Comment by eernstg** [comment]
It's about the binding of the new declared name to an object. That object may be mutable. For immutability of instances, check #125.

**Comment by yjbanov** [comment]
@mraleph I wouldn't be worried about JS semantics at all. Over there the `var` vs `let` vs `const` distinction is already too confusing to let that language inform what Dart should do.

@peldritch Swift does not make the rhs immutable. The following is valid Swift code:

```swift
class Mutable {
    var message: String = ""
}

let m = Mutable()
m.message = "What's up?"  // totally OK

// What it doesn't let you do is rebind the variable:
m = Mutable()  // ERROR
```

This is also the semantic we want in Dart.

**Comment by peldritch** [comment]
@yjbanov

Right, Swift `let` has a rhs effect for `struct` but not for `class`. So it's moot given Dart doesn't have structs.

**Comment by dlepex** [comment]
> I like `final` because it stands out.
Likes and dislikes are not very important for the sane language design, especially if you cannot justify them.

Most variables are "final", and only minority are vars. It is absolutely illogical for the common case to stand out! But for the rare case it may be a desirable property.

**Comment by Zhuinden** [comment]
It might make more sense to rename `var` to `mutable` so that people feel bad about writing it down.

**Comment by zoechi** [comment]
@dlepex 

> I like `final` because it stands out.

That depends on how much change is possible.

I'd prefer https://github.com/dart-lang/language/issues/136#issuecomment-447940385 any time.

> especially if you cannot justify them.

I think I did.

**Comment by munificent** [comment]
> Most variables are "final", and only minority are vars. It is absolutely illogical for the common case to stand out! But for the rare case it may be a desirable property.

There are a few separate concerns here that I think should be teased apart:

1. Most local variables are not re-assigned.

2. The author of the code wants to record their intent that the variable *cannot* be reassigned.

3. The reader of the code wants to see that intent reflected in the code.

The first point is observably empirically true. But it's not clear to me that the latter two are, at the level of local variables. There are a *lot* of intents that a programmer *could* write down for later programmers to know, and some of those could be mechanically verified by the analyzer. For example: 

* We could add an `undef` keyword to mark the end of a scope where a local variable should disappear. Most locals aren't used all the way to the end of a block, and most agree that the smaller the scope a variable has, the easier code is to understand.

* We could require users to add some marker when they refer to a variable declared outside of a lambda. Closures affect the lifetime of the object and are "less local", so maybe it would be helpful to force users to opt into it and see that it's happening.

* We could make array sizes part of their type, like Pascal does. That lets you statically avoid some array bounds errors.

* We could make a distinction between strings which can and cannot be empty. Lots of code considers it an error for an empty string to be passed, so why not check that mechanically?

With all of these, the question isn't "is the intent useful?" It's whether the intent useful *enough* to:

* Ask users to spend the mental effort to decide what their intent is and write it down.
* Ask all users to understand this feature of the language and be able to maintain code that uses it.
* Ask all users to spend time reading code that is more verbose because it expresses this intent.
* Ask users to go back and *change* their code when their intent changes. For example, if you later decide you *do* want to assign to some variable, you have to go back and turn the `final` (or whatever) into `var`.

For local variables, it's not clear to me that it's actually a net productivity gain to track which ones can be assigned and which can't. The variable is already local, so its scope is relatively small. It's often only a second's glance to tell if it is reassigned. Most editors will show all uses of a variable when you hover over it.

I *do* wish we had a shorter keyword for single assignment variables. (I pushed for `val` way back before Dart 1.0. Alas.) It would be particularly nice for fields and top level variables. But, for locals, I honestly don't think using `var` everywhere causes any measurable harm. It makes it easier when you do want to assign to them, and the fact that it isn't reassigned rarely improves the readability by any noticeable amount.

And, in general, I think it's important that we make a distinction between what the code the user wrote *happens* to do and what they *intend* it to be able to do. If I write a class and don't give it a private generative constructor, that doesn't necessarily mean I *intend* for users to subclass it. I may have simply not bothered to author any intent one way or the other. I think that's likely true of almost all uses of `var` for local variables.

**Comment by Hixie** [comment]
In Flutter we've been enforcing the use of `final` everywhere (except `for` loop variables and arguments, cc @pq) for a while, and I've found it really helpful to know immediately which variables are going to mutate and which are not. Surprisingly so, in fact. It's really confidence-building when reading new code if you can immediately know that a particular variable is not going to change, especially when there's multiple levels of complicated nested loops.

I wish we could opt-in (on a per-file or per-library basis) to making `final` the default, with `var Foo foo = ...` to declare a non-final field, removing `final` from the language.

**Comment by natebosch** [comment]
+1 on the value of seeing at declaration time whether a variable may be reassigned. I certainly _can_ scan down a function looking for assignments, but when I'm trying to wrap my head around new code `final` is a shortcut to understanding that is a huge benefit in the code that uses it.

I do not find it a drawback to need to go back and remove a `final` (or change a `let` to a `var`) when I add a new assignment to a variable - it's a reminder that my change isn't as shallow as I may have thought. And the same argument can be made for how easy this is - since the scope is small I shouldn't need to move far to do it.

**Comment by munificent** [comment]
One wrinkle with pushing towards single assignment locals by default is parameters. Even if we get `let` or `val` and encourage everyone to use it whenever possible, there's still the question of whether parameters should be single-assignment or not.

* If we think single-assignment is definitely better, than that's an argument that parameters should be implicitly final.

* On the other hand, an assignable parameter is strictly more useful than a single-assignment one. If you *do* want to assign to it, the current behavior already enables that without requiring you to opt in.

Scala, Kotlin, and Swift all treat parameters as single-assignment. They also all had ways to out of that which they then later deprecated and removed so that parameters are *only* single-assignment. That's a strong signal that it would be the right behavior for Dart if we moved to a shorter local variable keyword and encouraged everyone to use it. However, that would also be a massively breaking change. It's an automatically toolable one, but still. :-/

**Comment by munificent** [comment]
Oops, I may have misspoken about Scala. They may have always been single-assignment. Here is some discussion of the changes [in Swift](https://github.com/apple/swift-evolution/blob/master/proposals/0003-remove-var-parameters.md) and [Kotlin](https://blog.jetbrains.com/kotlin/2013/02/kotlin-m5-1/).

**Comment by Hixie** [comment]
Those seem like pretty compelling arguments, I'm sold.

**Comment by Hixie** [comment]
IDE's don't help when you're trying to understand the code in a YouTube video or on a slide at a conference or whatnot.

**Comment by zoechi** [comment]
And in GitHub. 
It also doesn't cause a warning if you assign to a variable that were declared final.

**Comment by davidmorgan** [comment]
Right. But people use `var` or `final` without type annotations, and people use unqualified imports instead of `show`, which is the same trade-off: it makes something visible only using the analyzer, not directly in the source. So it's not too shocking to lean on the analyzer.

**Comment by zoechi** [comment]
It's fine adding that but if possible not at the expense of other helpful features like `final`

**Comment by dlepex** [comment]
VSCode has a pretty nice plugin: https://github.com/siegebell/vsc-prettify-symbols-mode 

It can reduce any obnoxiously long keyword/operator into something really short. For instance "final" into "val", or even more radically to something like "◇", which is already quite close to Python/Go.

**Comment by MarcoLeongDev** [comment]
`let` and `var` does not introduce noise themselves:
```
var a = 1;
let b = 2;
var c = 3;

foobar(){
   print(a);
}
```

`final` as a keyword is visually noisy:
```
var a = 1;
final b = 2;
var c = 3;

foobar(){
   print(a);
}
```

This concept is mentioned by Kevlin Henney in ITT 2016, focusing on coding practices:
Code should be written so that visually it is understandable. A good indicator would be that the brain should not be struggling to classify the code into logical block with a glance at a page of code. Shown by the code above.

Base on this concept, `final` as a keyword is considered as bad design because the language is introducing the noise by default.

Going further into the discussion:
Your brain is mapping `var` and `let` as one block. If you require more information by focusing on the declaration "visual block", `var` and `let` starts and ends with different letters to given enough distinction when compare to the choice of `var` and `val`.


p.s. IDE/plugin should not be counted because they are not available when people review codes on their tablets. It also doesn't work with video and keynote slides as mentioned by Hixie in his/her comment.

**Comment by tatumizer** [comment]
If symbol ^ is accepted to denote immutability (like ^List\<int>) then the natural shortcut for "final int x` would be `int ^x`

**Comment by yjbanov** [comment]
@tatumizer I think before we start thinking of a specific syntax we should figure out the right immutability semantics we want. There are at least 3 popular use-cases that we need to decide on: shallow immutability (see e.g. https://github.com/dart-lang/language/issues/125#issuecomment-457132947 by @eernstg), non-shared deep immutability, x-isolate memory-shared immutability (@leafpetersen's [proposal](https://github.com/dart-lang/language/blob/master/working/0125-static-immutability/feature-specification.md)). It is not clear yet whether immutability will be a property of the type, a runtime property (I could swear I saw a proposal fly by, but can't find it anymore), or something modifier-driven. For the latter case your `^` could be used as a modifier.

**Comment by derekdreery** [comment]
Just to show prior art in other languages:

Rust uses `let` for immutable, `let mut` for mutable, and `const` for compile-time constant.

You probably don't need the `let` in class variables, so you could have

```dart

class MyClass {
    String immutableString; // this is equivalent to `final String immutableString` today.
    mut String mutableString; // This is equivalent to `String mutableString` today.
}
```

Here you're swapping the onus - things are immutable by default and you have to request mutability if you need it.

**Comment by minh98** [comment]
> `let` could be preferable to `val` or `:=` for immutable bindings:
> 
> `val` implies the rhs is a value object Kotlin style, not just a binding.
> `val` is tricky to distinguish from `var` when reading/eyeballing code.
> `let` is used in kernel.
> `let` is more pleasant to touch-type on english keyboards than `val`.
> `:=` is less touch-type friendly.
> `let result = when { } ` is aesthetic for expression/functional style programming.

i think i can typing 'val' faster than 'let' :D

**Comment by rrousselGit** [comment]
+1 for a setting to have `final` as default (a field in the `pubspec.yaml`?)

This is especially interesting for functions parameters.

I've realized some time ago that we could do:

```dart
void foo(final int bar) {
  bar = 42; // compile error
}
```

But specifying `final` for all parameters requires far too much boilerplate.

**Comment by minh98** [comment]
If `final` is set to default, the program will become confused, easier to read if using `let`

**Comment by psyanite** [comment]
In scala we use `var` and `final` with no issues 😁

**Comment by tatumizer** [comment]
Here's another outlandish (in a good sense!) idea.

Any replacement for "final" has to be significantly shorter than 5+1 chars (+1 is a space). I can think only about 3 numbers significantly smaller than 6, they are: 0, 1 and 2. As we will see shortly, all 3 of them are appear prominently in the following formalization.

Let's define the shortcuts:
`\\` - final
`\` - var

Examples:
```
\foo = ...;
\String foo = ...;
\\foo = ...;
\\String foo = ...;
```

For function parameters, "final" will be a default, so we will need explicit `\` to make it non-final.
The convention makes declarations more visible, and - due to a better visibility - we can support declarations as part of expressions, like:
```
\x = (\y = a +b) + (\z = d + e); // introduces also y and z
for (\i = 0; i<10; i++) {...} //  \i instead of "int i"
```

**Comment by Zhuinden** [comment]
>  another outlandish (in a good sense!) idea.
>
> `\\` for immutable 
>
> `\` for mutable

I'm a bit confused, it still takes extra effort to make something immutable than mutable, and the clarity of `var` and `let/val` is lost in comparison.

**Comment by tatumizer** [comment]
@Zhuinden : (it's not for mutable/immutable, but for final/non-final). Counting to 1 vs 2 has never been a problem in `-x` vs `--x`, `foo.x=1` vs `foo..x=1` etc. The  advantage of outlandish declaration notation is not only brevity and mathiness, but also the possibility of inline declarations inside expressions. The idea can be pushed a notch further, by supporting "where" notation:
```
\x= \y + \z where y = a+b, z=d+e;
```
Expressions like this can help to flatten the towers in Flutter object literals, thus eliminating lots of syntactic noise.

**Comment by psyanite** [comment]
What if we rename `final` to `fin`?

**Comment by erf** [comment]
I like `let`. `let` and `var` would be consistent with the Swift language - which is awesome. 

Now just add pattern matching, enum with values, tuples and guards 😁

**Comment by tatumizer** [comment]
Swift doesn't support anything like `late final` (in swift's terms, that would be `lazy let`). See [this exchange](https://stackoverflow.com/questions/27886024/lazy-var-vs-let)
Though it's really hard to make a comparison because `lazy` in swift does not have the same meaning as `late` in dart - in any case, be it late or lazy, the expressions `late let` or `lazy let` both sound like gibberish to me. (I can speculate that swift doesn't have `lazy let` for that exact reason, but I can be wrong).

I don't think renaming `final` will do much good, but if we are concerned with the length of the word then (IMO) we have to look for a single symbol, which mentally could be pronounced as `final` while reading the code.

**Comment by psyanite** [comment]
In Scala we use `lazy val customers = customerSvc.getAllCustomers()`
We have `var` (reassignable variable) and `val` (non-reassignable value).

**Comment by lukepighetti** [comment]
What about allowing people to use `final` or `fin`?

**Comment by cedvdb** [comment]
If you make final the default as opt-in, most people would opt-in, and it wouldn't break existing code. Libraries would slowly migrate a la null safe.  I didn't see any counter arguments for this in the thread

**Comment by willhaslett** [comment]
I'm confused by the discussions above linking `final` to the concept of mutability. I'm relatively new to Dart, but doesn't `final` simply mean that the declared variable cannot later be assigned to a different object, not that the object to which it refers cannot be mutated? For classes with no mutable public properties, like `String`, `int`, etc., this may be effectively the same thing, but this works, where `Frog`s default constructor exposes a `color` property, right?
```
final Frog frog = Frog();
frog.color = 'pink';
```
I mention this because it seems important not to confuse that `final` does with mutability. That exact confusion cost me some growing pains when getting started with the language, and it's a reason I'm not a fan of `final`, in addition to it taking up space.

**Comment by lukepighetti** [comment]
What's super interesting is people who write Flutter tend to use `final` to communicate that which should never change regardless of immutability, while people who write in other Dart projects tend to use `var` for locals regardless of if it should change or not

**Comment by lukepighetti** [comment]
I actually disagree that `final` is too long at this point.

```
const
final
var
```

const and final are grouped visually by the character length

**Comment by munificent** [comment]
> Why this divergence has occurred is not clear to me

I think a lot of Flutter users adopt the style of the Flutter framework itself, which is quite different from the prior idiomatic Dart style.

**Comment by cedvdb** [comment]
> Correct. After much discussion many of us have settled on always using `var` for locals; it's why I added the lint [unnecessary_final](https://dart-lang.github.io/linter/lints/unnecessary_final.html).

using final shouldn't be used as safe guard but as to communicate intent imo. That intent being, this variable shouldn't be reassigned.

**Comment by lukepighetti** [comment]
> > Why this divergence has occurred is not clear to me
> 
> I think a lot of Flutter users adopt the style of the Flutter framework itself, which is quite different from the prior idiomatic Dart style.

Or to make it clear which entities in a state class are mutable with `setState()`, and which entities are constant in nature.

**Comment by rrousselGit** [comment]
final local variables are not just to match flutter's styleguide

It is an important practice to avoid reassigning local variables and prefer creating a new variable when possible.
It's a common readability nightmare in legacy code where a variable is changed in the middle of a 200 lines method.

**Comment by venkatd** [comment]
We default to using final for all variables. Both for classes and local variables. There's a reduced mental burden when reading code knowing that something can't be reassigned.

**Comment by Hixie** [comment]
`dart fix` makes it even more plausible that we could literally flip the language around here, making everything `final` by default (and remove `final` from the language), and if you want to make something assignable, it has to be marked `var`.

```dart
class Foo {
  Foo(this.id);
  var String value = 'mutable';
  int id; // immutable
  List<Foo> children = <Foo>[]; // list contents are mutable, `children` member is not

  int test(int immutable, var int mutable) {
    for (var int index = 0; index < immutable; index +=1 ) { // ok maybe we should imply `var` here...
      print(index);
    }
    for (Foo child in children) { // `child` is immutable
      mutable += 1;
    }
    return mutable;
  }
```

It might be worth talking about changing the semantics of the index in a C-style for loop, by the way. Right now they're entirely mutable. Would be good to make them immutable *except in the third clause of the for loop declaration* by default (and then `var` could be used to return to today's behaviour).

Anyway the point is that `dart fix` would let us do this entirely mechanically as a language upgrade, just like we did the null safety migration.

**Comment by lrhn** [comment]
Removing `final` means that you have to write a type on all final variables, which can probably get annoying (unless we also introduce the `:=` final assignment, but then we don't actually need to remove `final`).
In general, we do want to have ways to infer types of variables with initializer expressions, even if final.

The full "final by default" proposal that I started at some point had: 
* All variables final by default, unless you write `var`.
* For `for`-`in` you don't need `var` or a type: `for (x in ...)` is implicitly `final x`. (Ditto for parameters). You can't reuse an existing variable any more (so `int? last; for (last in foo); print(last);` won't work, you have to rewrite as `int? last; for ($tmp in foo) { last = foo; } print(last);`. I hope nobody knows it ever could work.
* (The `catch (e, s)` variables are already final)

The `for(;;)` loop is tricky in that it's several idioms in one, and not all of them fit.
I guess a final variable declaration can be tweaked into a variable variable declaration that cannot be modified outside of the increment part. That's no weirder than what's already going on with variables in for loops.

**Comment by Hixie** [comment]
I'd be fine with leaving `final` for that specific case of declaring a variable without a type (same as `var` today).

**Comment by leonsenft** [comment]
> I'd be fine with leaving `final` for that specific case of declaring a variable without a type (same as `var` today).

I think this misses the point, which is that many want the convenience and shorter notation of `var`, with the benefits of `final` by default.

My understanding is that Flutter prefers LHS types , but google3 enforces [omit_local_type_variables](https://dart-lang.github.io/linter/lints/omit_local_variable_types.html), so ideally we could land on a solution that satisfies both parties.

**Comment by Hixie** [comment]
We can spell `final` with fewer letters for that case, like `val` or `let`, sure.

**Comment by tatumizer** [comment]
Another possibility: while omitting explicit type (relying on type inference), declare the variable as "auto" type, e.g.
`auto foo = 42;`

**Comment by tomwyr** [comment]
I'm not sure if there's any advantage in inventing yet another way of declaring immutable object references. As @Hixie mentioned, either `val` or `let`  would be probably least confusing since both are already present in widely used programming languages like Kotlin, Scala, Rust, Swift.

**Comment by tatumizer** [comment]
Will you allow the form of `let` with the explicit type, like `let int x=42;` ?

**Comment by tbm98** [comment]
I think `final` is fine and easiest way to declared a `final auto type`

**Comment by cedvdb** [comment]
I think @Hixie was misunderstood. He suggested making values final by default. IE have those three be equal:

```dart
final String a = 'str';
String a = 'str';
final a = 'str';
```

**Comment by Hixie** [comment]
my proposal was intended to match the current status quo, with just the default behaviour flipped. Today you can't say `var String a = 'str'`.

**Comment by cedvdb** [comment]
> my proposal was intended to match the current status quo, with just the default behaviour flipped. Today you can't say var String a = 'str'.

An analyzer should check the percentage of `final` vs `var` to see if that makes sens. I personally would like it.

**Comment by tatumizer** [comment]
If you write `final a = 'str';`, you are emphasizing the finality of a. Which is a bit odd if we make "final" a default.

Also, this opens a way to using two styles of declarations: one with the explicit final, another - with final by default:
```dart
final int x=42;
final x=42;
```
Which is not very good. That's (probably) why "var" doesn't support explicit type (`var int x=42;` is an error).

Yet another (semantic) problem is that the word "final" in the declaration `final x=42;` compensates for the absence of explicit *type*. But these two notions ("final" and "type") are totally different. It's like making the word "blue" a substitute for "moon". That's where the idea of "auto" comes from. "auto" means "auto-type".

**Comment by cedvdb** [comment]
> Also, this opens a way to using two styles of declarations: one with the explicit final, another - with final by default:
> final int x=42;
> final x=42;
> Which is not very good

This is already how it works?

 >  But these two notions ("final" and "type") are totally different.
 
The point is that writing `final String x` or `final x` is supposedly more idiomatic than either `var x` or `String x`.  While the two first expressions are used way more than the two later, both are more verbose. Switching the default would give:

```dart
final x = 'final';
String x = 'final';
var x = 'final';
var String x = 'final';
```

**Comment by tatumizer** [comment]
> This is already how it works?
Sorry, I meant
```dart
final int x=42;
int x=42;
```
Fixed it.

**Comment by willhaslett** [comment]
> > I'd be fine with leaving `final` for that specific case of declaring a variable without a type (same as `var` today).
> 
> I think this misses the point, which is that many want the convenience and shorter notation of `var`, with the benefits of `final` by default.
> 
> My understanding is that Flutter prefers LHS types , but google3 enforces [omit_local_variable_types](https://dart-lang.github.io/linter/lints/omit_local_variable_types.html), so ideally we could land on a solution that satisfies both parties.

@leonsenft, tangential to your point, but I find it odd that `omit_local_variable_types` uses the word "usually" in its rationale. Is it that the worst, rare case is that type inference is doable but expensive? If it can ever fail...

**Comment by leonsenft** [comment]
@willhaslett 

> Is it that the worst, rare case is that type inference is doable but expensive?

No, it's because sometimes inference will give you a type you don't want - for example one that's _too_ specific:

```dart
abstract class Node {
  Node? get parent;

  Node? ancestorThat(bool Function(Node) matches) {
    var current = this; // Inferred as `Node`.
    while (current != null) {
      if (matches(current)) {
        break;
      }
      current = current.parent; // Error; can't assign `Node?` to `Node`.
    }
    return current;
  }
}
```

One solution here is to widen the type of current with a type annotation to allow null:

```dart
    Node? current = this;
```

**Comment by ykmnkmi** [comment]
@leonsenft same cases, sometimes I'm looking for feature like this

```dart
var? current = this;
```

**Comment by tatumizer** [comment]
Here's a rather radical syntax idea for "final by default":

- Introduce "let" (in addition to final and var for the time being, but mark both as "to be deprecated").
- "let" will alow to declare both final and non-final, but there's more hassle to declare non-final
- if explicit type is used, there's no need to write "let"
```dart
let x=42; // final
let \x=42; // non-final is declared with backslash
int x=42; // final
int \x=42; // non-final
print(x); // use it with no back slash
```
(Maybe backslash is not the best symbol, but - at least on US keyboards (not sure about others) - it can be hit without pressing SHIFT).

**Comment by Levi-Lesches** [comment]
I don't think we should make non-final look more confusing than it is today. Final by default, meaning a few extra letters for non-final, is fine, but I don't see the value in a punctuation symbol over `var` or `String`.

**Comment by Wdestroier** [comment]
I'd say final is fine... It is only annoying in Java, because you need to write `final var x = ...;`
Making the `var` keyword final is strange, it would define a variable that isn't variable!
And the programmer would need another keyword to define a variable that is variable.
`val` vs `var` is too easy to miss when reading the code and `let` is kind of meaningless...

**Comment by mateusfccp** [comment]
>Making the `var` keyword final is strange, it would define a variable that isn't variable!

This isn't true... A `final` variable is not immutable, it's only not reassignable. You can still, for instance, make a `final a = <int>[]` and then call `a.add(0)`. The `final` variable will vary.

**Comment by Wdestroier** [comment]
> The `final` variable will vary.

The variable won't vary, what will vary is the class instance field value.

```dart
class Text { // Text is a class that explicitly extends
  // the Object class, thus it is an object.
  String value; // value is a field, String is the field type.

  Text(this.value);
}

void main() {
  // name is a variable.
  var name;
  // new Text is a class instance
  name = Text('Gates');
  
  // The Text class instance is not immutable.
  // Because it contains fields that can have their value changed.
  (name as Text).value = 'Steve';
}
```

Take as example the following line, which is valid code in Dart:
```dart
var name;
```
This line declares a variable reference called "name". This variable may reference a mutable or an immutable object instance.
Example:
```dart
var name = const Text('John'); // the var keyword is not related to what it references.
```

If the `var` keyword was final, then the line above would declare a _final reference to an immutable object_. Nothing would be variable at all.

**Comment by lrhn** [comment]
Depending on context, the word "immutable" often carries connotations of *deep* immutability.
I think it's fair to point out that final variables are not that kind of immutable. They're what we have elsewhere called "unmodifiable". A `const` variable is deeply immutable, on the other hand. (Well, structurally, not necessarily logically, if it pretends to have mutable state using expandos).

That said, we're talking about Dart `final` variables here, and their meaning is well defined.

**Comment by Squinnb** [comment]
It would be nice even just to shorten it to fin.

**Comment by hepin1989** [comment]
val？

**Comment by cedvdb** [comment]
val is too close to var lexically.

I don't think it needs to be shortened personally, it's quite expressive as is. I'd rather have it the default as I tend to want to explicitly say that something needs to be reassigned but adding final to parameters is too verbose. The `final String houseName`  instead of `final houseName` for flutter lints does not make sense to me so I just use `final name` and it's short enough.

> tangential to your point, but I find it odd that `omit_local_variable_types` uses the word "usually" in its rationale. Is it that the worst, rare case is that type inference is doable but expensive? If it can ever fail...

Maybe for cases where you do want another type. eg `final Pattern x = '...'` instead of `final x = '...''`, I'm not sure if that ever makes sense? The second example here does not make sense either https://dart-lang.github.io/linter/lints/omit_local_variable_types.html but it seems to indicate that's why.

**Comment by gintominto5329** [comment]
Also, I have a question, 

Why cannot, `const`, and `final`; be made automatic, as the analyzer knows, that
- the called constructor is `const` or not, and
- the re-assignment of variables,

thanks again,

**Comment by lrhn** [comment]
Too late for Dart 3 to add another major feature. 

I too would vote for "final by default", as I've suggest up-thread, but it's not a small change. It requires migrating all code in the world. (Again!)
It should be completely automatable, at least, unlike null safety, which also suggests that it doesn't actually *change* anything except the default. There is no new power, no new protection, like there was with null safety. That lowers the benefit in the cost/benefit ratio.

Making `final` automatic when a variable is not assigned to is ... basically status quo. If you add an assignment, the variable would stop being `final`. Whether the variable is final or not should not have any effect if there is no assignment. You can't distinguish "implicitly `final` if not reassigned" from just plain `var`.

Making `const` automatic when calling a `const` constructor with constant arguments is possible, but has some sharp edges. First of all, it doesn't matter for anything except canonicalization of constant values. There is no discernible semantic difference between a `new ConstConstructor(constArgument)` and `const ConstConstructor(constArgument)` other than the identity of the object - the latter is identical to other similar `const` invocations. The real difference is memory consumption or churn, if you create a lot of `new` objects.

Also, whether `const` can be omitted gets tricky around collection literals.
```dart
var c1 = ConstantConstructor(const []); // Would be implicitly `const`?
var c2 = const ConstantConstructor([]); // Is explicitly `const`.
var c3 = ConstantConstructor([]); // Would not be implicitly `const`, because the argument is not constant.
```
It would be easy to take a big constant structure like `var c = const Something(Big(Long(Deep(42, [], "banana")));` and think you could remove the `const`. Or have a big structure that is implicitly const, forget that it's not a const context, and add an empty list like `[]` somewhere, and suddenly the entire structure becomes non-constant.

And you *wouldn't notice*, because it still runs, it just allocates a lot more, mostly indistinguishable, objects.

We considered this at some point, and decided that the risks were too big, and it's better to make you *always* write at least one `const` when you want to create a constant object. It's worse if you must write it sometimes, and don't have to in other cases, and the only difference can be a single two-character change deep in a large structure.

**Comment by whoizit** [comment]
final is too long, but I hate :=

**Comment by myConsciousness** [comment]
Maybe this topic has already changed topics, but based on the title I am writing my opinion.

When I read this title, I thought it would be nice if the variables were final from the beginning, but then I would have to write the type without fail, forcing even more strange declarations.

On the other hand, I would not like to see strange syntax added. (`:=`? Please don't!) I like Dart's `final` declaration very reasonably and have no complaints at all.

**Comment by modulovalue** [comment]
If we take this issue to the extreme, I think an argument could be made that other keywords like `extension`, `required` and `Function` are also too long. We could abbreviate extension to `ext` (6 characters saved) required to `req` (5 characters saved) and `Function` to `<-` (6 characters saved). When should we stop? 


Here's perhaps an unpopular opinion: I think that the approach that Dart takes **is far superior** to all the alternatives proposed in this issue.

* final clearly states the intended effect that `this variable can't be assigned to again` and `final` is semantically much closer to that idea than `val`, `let` or `:=`.

* mutable by default allows new developers to be productive and make progress much quicker because they are not forced to use concepts that they really don't need.

* I think that syntax should be optimized for reading and not for writing because code is read more often than it is written. I agree that typing two extra characters can be annoying, but when reading code, the two extra characters are really helpful in finding out what's mutable and what's not.

---

I agree that the ability to write those keywords quicker would make developers slightly more productive when writing code. But if the goal is to reduce keystrokes, then I'd recommend the pattern that is known as the hyper key on macos ([hyper key via karabiner elements](https://holmberg.io/hyper-key/) / [hyper key via better touch tool](https://docs.folivora.ai/docs/1004_hyper_key.html)). It allows you to reassign caps lock to a new modifier. That way you can get your keystrokes down to 2 keys (e.g. `hyper+f` to type `final`, `hyper+r` to type `required`, `hyper+n` to type Function and so on).

**Comment by ds84182** [comment]
Not a fan of `:`. Not very searchable, and variable shadowing across statement blocks would have to become an error to avoid very obvious footguns.

`val` is better suited for this. `let` may conflict with let expressions.

**Comment by johnpyp** [comment]
I see those two CLs are currently abandoned. What are the next steps for this?

**Comment by gochev** [comment]
Adding multiple keywords that so almost the same or actually the same is so so bad. It's like DRY on language level :)
